docs/source/external_exporters.rst

```diff
@@ -0,0 +1,156 @@
+.. _external_exporter
+
+Using external exporters
+========================
+
+.. versionadded:: 4.2
+
+    You can now use the ``--to`` flag to use custom export formats defined in other
+    libraries.
+
+
+The command-line syntax to run the ``nbconvert`` script is::
+
+  $ jupyter nbconvert --to FORMAT notebook.ipynb
+
+This will convert the Jupyter document file ``notebook.ipynb`` into the output
+format designated by the ``FORMAT`` string as explained below.
+
+A few built-in formats are available by default: `html`, `pdf`,
+`script`, `latex`. Each of these has its own _exporter_ with many configuration
+options that can be extended. Having the option to point to a different _exporter_ 
+allows authors to create their own fully customized templates. 
+
+A custom _exporter_ must be a globally importable python object. We recommend that
+these be distributed as python libraries.
+
+To export using an exporter from an external library, use the `fully qualified`
+name of this exporter on the command line as the argument of the ``--to`` flag::
```

Comment:
Ohh, cool. Some questions:

* Does it find/load regular setuptools entry points?
* How does one declare the entrypoints?
* Who needs to install it? Consumers of entrypoints? Providers of
entrypoints? Both?
* For us to rely on something like this in our core stack, I think it would
need to be pretty stable and battle tested. How far aware is it from that?

Because this type of thing is starting to come up in many different parts
of our projects, I think we should figure out a robust consistent approach
that all our projects can use. Having different projects use different
approaches will only confuse users. Maybe we want to bring this up on the
list or as an enchancemnt proposal?

Just to clarify, I have "re-invented" entry points in this PR as well:

https://github.com/jupyter/notebook/pull/879

which is targeted for a quick 4.2 release of jupyter/notebook so it would
be good to figure this stuff out on a pretty short time frame.




On Wed, Jan 6, 2016 at 9:57 AM, Thomas Kluyver <notifications@github.com>
wrote:

> In docs/source/external_exporters.rst
> <https://github.com/jupyter/nbconvert/pull/141#discussion_r48987153>:
>
> > +
> > +  $ jupyter nbconvert --to FORMAT notebook.ipynb
> > +
> > +This will convert the Jupyter document file ``notebook.ipynb`` into the output
> > +format designated by the ``FORMAT`` string as explained below.
> > +
> > +A few built-in formats are available by default: `html`, `pdf`,
> > +`script`, `latex`. Each of these has its own _exporter_ with many configuration
> > +options that can be extended. Having the option to point to a different _exporter_
> > +allows authors to create their own fully customized templates.
> > +
> > +A custom _exporter_ must be a globally importable python object. We recommend that
> > +these be distributed as python libraries.
> > +
> > +To export using an exporter from an external library, use the `fully qualified`
> > +name of this exporter on the command line as the argument of the ``--to`` flag::
>
> The entrypoints <https://pypi.python.org/pypi/entrypoints> module is my
> attempt to free the technology from setuptools - it's a completely
> standalone module that can find and load entry points, so we wouldn't need
> nbconvert to depend on setuptools. It's not battle-tested yet, but I think
> it's simple enough that we can iron out any bugs that come up.
>
> Third parties who wanted to offer nbconvert exporters would need to either
> use setuptools (which almost all Python packages now do anyway), or package
> with something like flit.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jupyter/nbconvert/pull/141/files#r48987153>.
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com